File: Questions 161 - 170/169_majorityElement.py
# check out Boyer-Moore Voting Algorithm
from typing import List

def majorityElement(nums: List[int]) -> int:
    
    if(len(nums)==1):
        return nums[0]
    
    count = dict()
    max_num = -1
    max_num_count = -1
    
    for num in nums:
        if num not in count:
            count[num] = 1
        else:
            count[num] += 1
            if(count[num]>max_num_count):        
                max_num_count = count[num]
                max_num = num
    
    return max_num

if __name__ == "__main__":
    n = list(input("Enter list of elements: ").split())
    nums = list(map(int, n))

    res = majorityElement(nums)

    print(f"Majority element is: {res}")


    

File: Questions 101 - 200/Questions 161 - 170/169_majorityElement.cpp
#include<iostream>
#include<vector>
using namespace std;

int majorityElement(vector<int>& nums) {
    // moore's voting algorithm
    int count = 0, majority;
    
    for(int x: nums) {
        if(!count){
            majority = x;
        }
        count += (majority == x) ? +1 : -1;
    }
    
    return majority;
}

int main() {
    vector<int> arr;
    int n, ele;

    cout<<"Enter array size: ";
    cin>>n;

    for(int i=0; i<n; i++) {
        cout<<"Enter element "<<i+1<<" : ";
        cin>>ele;
        arr.push_back(ele);
    }

    cout<<"Majority Element: "<<majorityElement(arr);
    
}